Don't resize outside timeline boundary

ALevansSamsung opened the pull request:
#### What it does

This PR is made in response to [this issue](https://github.com/android-graphics/sokatoa/issues/1286) regarding how track resizing handles the bottom track group.  It was brought up that the resizing continued even when the mouse left the scope of the timeline and/or application.  This PR stops resizing a track or track group if the mouse leaves the boundaries of the timeline

ALevansSamsung committed:
Don't resize outside timeline boundary

ALevansSamsung commented:
@jcortell68 @dfriederich I'm not sure if this is closer to a solution for the "Resizing using the bottom border of the last track/group is buggy" issue, but I'd like to get opinions and feedback on this functionality.  From my opinion killing resizing after the mouse leaves the timeline could cause situations where users may not be able to resize tracks or track groups as large as they want.

jcortell68 commented:
@ALevansSamsung For the sake of those of us not working on Linux (and thus not able to build perfett ui), could you please post the `.js` versions of the files you modified. I can plug them into my environment and try the new behavior.

ALevansSamsung commented:
[built files.zip](https://github.com/user-attachments/files/15994315/built.files.zip)

ALevansSamsung commented:
track_panel.js
```
"use strict";
// Copyright (C) 2018 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrackPanel = exports.TrackButton = exports.TrackContent = void 0;
const tslib_1 = require("tslib");
const color_convert_1 = require("color-convert");
const mithril_1 = tslib_1.__importDefault(require("mithril"));
const actions_1 = require("../common/actions");
const css_constants_1 = require("./css_constants");
const globals_1 = require("./globals");
const gridline_helper_1 = require("./gridline_helper");
const icons_1 = require("./icons");
const panel_1 = require("./panel");
const scroll_helper_1 = require("./scroll_helper");
const track_registry_1 = require("./track_registry");
const vertical_line_helper_1 = require("./vertical_line_helper");
const vsync_helper_1 = require("./vsync_helper");
const state_1 = require("../common/state");
function isPinned(id) {
    return globals_1.globals.state.pinnedTracks.indexOf(id) !== -1;
}
function isSelected(id) {
    const selection = globals_1.globals.state.currentSelection;
    if (selection === null || selection.kind !== 'AREA')
        return false;
    const selectedArea = globals_1.globals.state.areas[selection.areaId];
    return selectedArea.tracks.includes(id);
}
class TrackShell {
    constructor() {
        // Set to true when we click down and drag the
        this.dragging = false;
        this.dropping = undefined;
        this.resize = (e) => {
            e.stopPropagation();
            e.preventDefault();
            if (!this.attrs) {
                return;
            }
            if (e.currentTarget instanceof HTMLElement) {
                const pageElement = e.currentTarget.closest('div.page');
                if (pageElement && pageElement instanceof HTMLDivElement) {
                    let trackHeight = this.attrs.track.getHeight();
                    let mouseY = e.clientY;
                    const mouseMoveEvent = (evMove) => {
                        evMove.preventDefault();
                        if (!this.attrs) {
                            return;
                        }
                        let movementY = evMove.clientY - mouseY;
                        if (this.attrs.pinnedCopy !== true &&
                            isPinned(this.attrs.trackState.id)) {
                            movementY /= 2;
                        }
                        trackHeight += movementY;
                        mouseY = evMove.clientY;
                        if (this.attrs && this.defaultHeight) {
                            const newMultiplier = trackHeight / this.defaultHeight;
                            if (newMultiplier < 1) {
                                this.attrs.trackState.scaleFactor = 1;
                            }
                            else {
                                this.attrs.trackState.scaleFactor = newMultiplier;
                            }
                            globals_1.globals.rafScheduler.scheduleFullRedraw();
                        }
                    };
                    const mouseReturnEvent = () => {
                        pageElement.addEventListener('mousemove', mouseMoveEvent);
                        pageElement.removeEventListener('mouseenter', mouseReturnEvent);
                    };
                    const mouseLeaveEvent = () => {
                        pageElement.removeEventListener('mousemove', mouseMoveEvent);
                        pageElement.addEventListener('mouseenter', mouseReturnEvent);
                    };
                    const mouseUpEvent = () => {
                        pageElement.removeEventListener('mousemove', mouseMoveEvent);
                        pageElement.removeEventListener('mouseup', mouseUpEvent);
                        pageElement.removeEventListener('mouseenter', mouseReturnEvent);
                        pageElement.removeEventListener('mouseleave', mouseLeaveEvent);
                    };
                    pageElement.addEventListener('mousemove', mouseMoveEvent);
                    pageElement.addEventListener('mouseleave', mouseLeaveEvent);
                    pageElement.addEventListener('mouseup', mouseUpEvent);
                    pageElement.removeEventListener('mousedown', this.resize);
                }
            }
        };
    }
    oninit(vnode) {
        this.attrs = vnode.attrs;
        if (this.attrs) {
            this.defaultHeight =
                this.attrs.track.getHeight() / this.attrs.trackState.scaleFactor;
        }
    }
    view({ attrs }) {
        // The shell should be highlighted if the current search result is inside
        // this track.
        let highlightClass = '';
        const searchIndex = globals_1.globals.state.searchIndex;
        if (searchIndex !== -1) {
            const trackId = globals_1.globals.currentSearchResults.trackIds[searchIndex];
            if (trackId === attrs.trackState.id) {
                highlightClass = 'flash';
            }
        }
        const depth = (attrs.trackState.trackGroup === state_1.SCROLLING_TRACK_GROUP ?
            0 :
            (0, state_1.getContainingTrackIds)(globals_1.globals.state, attrs.trackState.id)?.length ?? 0) +
            1;
        const trackTitle = attrs.trackState.title ?? attrs.trackState.name;
        const titleStyling = {};
        titleStyling.marginLeft = `${depth / 2}rem`;
        const dragClass = this.dragging ? `drag` : '';
        const dropClass = this.dropping ? `drop-${this.dropping}` : '';
        return (0, mithril_1.default)(`.track-shell[draggable=true]`, {
            class: `${highlightClass} ${dragClass} ${dropClass}`,
            ondragstart: this.ondragstart.bind(this),
            ondragenter: (e) => {
                e.preventDefault();
                e.stopPropagation();
            },
            ondragend: this.ondragend.bind(this),
            ondragover: this.ondragover.bind(this),
            ondragleave: this.ondragleave.bind(this),
            ondrop: this.ondrop.bind(this),
            onmousemove: this.onmousemove.bind(this),
            onmouseleave: this.onmouseleave.bind(this),
        }, (0, mithril_1.default)('h1', {
            title: attrs.trackState.description,
            style: titleStyling,
        }, trackTitle, ('namespace' in attrs.trackState.config) &&
            (0, mithril_1.default)('span.chip', 'metric')), (0, mithril_1.default)('.track-buttons', ...this.getTrackShellButtons(attrs), attrs.track.getContextMenu(), (0, mithril_1.default)(TrackButton, {
            action: (e) => {
                // Scroll timeline by height of toggledPinnedTrack
                const toBePinned = !globals_1.globals.state.pinnedTracks.includes(attrs.trackState.id);
                if (e.currentTarget && e.currentTarget instanceof Element) {
                    const trackShell = e.currentTarget.closest('.track-shell');
                    if (trackShell) {
                        let toScroll = trackShell.clientHeight;
                        if (!toBePinned) {
                            toScroll *= -1;
                        }
                        const parentScrollPanel = trackShell.closest('.scrolling-panel-container');
                        if (parentScrollPanel) {
                            parentScrollPanel.scroll(0, parentScrollPanel.scrollTop + toScroll);
                        }
                    }
                }
                globals_1.globals.dispatch(actions_1.Actions.toggleTrackPinned({ trackId: attrs.trackState.id }));
            },
            i: icons_1.PIN,
            filledIcon: isPinned(attrs.trackState.id),
            tooltip: isPinned(attrs.trackState.id) ? 'Unpin' : 'Pin to top',
            showButton: isPinned(attrs.trackState.id),
            fullHeight: true,
        }), globals_1.globals.state.currentSelection !== null &&
            globals_1.globals.state.currentSelection.kind === 'AREA' ?
            (0, mithril_1.default)(TrackButton, {
                action: (e) => {
                    globals_1.globals.dispatch(actions_1.Actions.toggleTrackSelection({ id: attrs.trackState.id, isTrackGroup: false }));
                    e.stopPropagation();
                },
                i: isSelected(attrs.trackState.id) ? icons_1.CHECKBOX : icons_1.BLANK_CHECKBOX,
                tooltip: isSelected(attrs.trackState.id) ?
                    'Exclude track from area selection' :
                    'Include track in area selection',
                showButton: true,
            }) :
            ''));
    }
    onmousemove(e) {
        if (this.attrs?.track.supportsResizing) {
            if (e.currentTarget instanceof HTMLElement &&
                e.pageY - e.currentTarget.getBoundingClientRect().top >=
                    e.currentTarget.clientHeight - 5) {
                const pageElement = e.currentTarget.closest('div.page');
                pageElement?.addEventListener('mousedown', this.resize);
                e.currentTarget.style.cursor = 'row-resize';
                return;
            }
            else if (e.currentTarget instanceof HTMLElement) {
                const pageElement = e.currentTarget.closest('div.page');
                pageElement?.removeEventListener('mousedown', this.resize);
                e.currentTarget.style.cursor = 'unset';
            }
        }
    }
    onmouseleave(e) {
        if (this.attrs?.track.supportsResizing &&
            e.currentTarget instanceof HTMLElement) {
            e.currentTarget.style.cursor = 'unset';
            const pageElement = e.currentTarget.closest('div.page');
            pageElement?.removeEventListener('mousedown', this.resize);
        }
    }
    ondragstart(e) {
        const dataTransfer = e.dataTransfer;
        if (dataTransfer === null)
            return;
        this.dragging = true;
        e.stopPropagation();
        globals_1.globals.rafScheduler.scheduleFullRedraw();
        dataTransfer.effectAllowed = 'move';
        dataTransfer.setData('perfetto/track/' + this.attrs.trackState.id, `${this.attrs.trackState.id}`);
        dataTransfer.setDragImage(new Image(), 0, 0);
    }
    ondragend() {
        this.dragging = false;
        globals_1.globals.rafScheduler.scheduleFullRedraw();
    }
    ondragover(e) {
        if (this.dragging)
            return;
        if (!(e.target instanceof HTMLElement))
            return;
        const dataTransfer = e.dataTransfer;
        if (dataTransfer === null)
            return;
        const dataType = dataTransfer.types.find((dataType) => {
            return dataType.startsWith('perfetto/track/');
        });
        if (!dataType)
            return;
        const trackLikeId = dataType.split('/').pop();
        if (!trackLikeId)
            return;
        e.stopPropagation();
        dataTransfer.dropEffect = 'move';
        e.preventDefault();
        // Test if id has same parent as current
        // If not do not set this.dropping
        const trackLike = globals_1.globals.state.trackGroups[trackLikeId] ??
            globals_1.globals.state.tracks[trackLikeId];
        if (('trackGroup' in trackLike && this.attrs.trackState.trackGroup === trackLike.trackGroup) ||
            'parentGroup' in trackLike && this.attrs.trackState.trackGroup === trackLike.parentGroup) {
            // Apply some hysteresis to the drop logic so that the lightened border
            // changes only when we get close enough to the border.
            if (e.offsetY < e.target.scrollHeight / 3) {
                this.dropping = 'before';
            }
            else if (e.offsetY > e.target.scrollHeight / 3 * 2) {
                this.dropping = 'after';
            }
        }
        globals_1.globals.rafScheduler.scheduleFullRedraw();
    }
    ondragleave() {
        this.dropping = undefined;
        globals_1.globals.rafScheduler.scheduleFullRedraw();
    }
    ondrop(e) {
        if (this.dropping === undefined)
            return;
        const dataTransfer = e.dataTransfer;
        if (dataTransfer === null)
            return;
        globals_1.globals.rafScheduler.scheduleFullRedraw();
        const dataType = dataTransfer.types.find((dataType) => {
            return dataType.startsWith('perfetto/track/');
        });
        if (!dataType)
            return;
        const srcId = dataType.split('/').pop();
        if (!srcId)
            return;
        const dstId = this.attrs.trackState.id;
        globals_1.globals.dispatch(actions_1.Actions.moveTrack({ srcId, op: this.dropping, dstId }));
        this.dropping = undefined;
    }
    getTrackShellButtons(attrs) {
        const result = [...attrs.track.getTrackShellButtons()];
        result.push((0, mithril_1.default)(TrackButton, {
            action: () => globals_1.globals.dispatch(actions_1.Actions.removeTrack({ trackId: attrs.trackState.id })),
            i: 'hide',
            tooltip: 'Hide track',
            showButton: false,
            fullHeight: true,
            disabled: !this.canDeleteTrack(attrs.trackState),
        }));
        return result;
    }
    // We cannot delete a track while it is loading, otherwise
    // we'll try to read data from tables that have been dropped.
    // We assume it may be loading if its engine is busy.
    canDeleteTrack(trackState) {
        const engine = globals_1.globals.engines.get(trackState.engineId);
        return !engine || !engine.hasDataPending;
    }
}
class TrackContent {
    constructor() {
        this.selectionOccurred = false;
    }
    view(node) {
        const attrs = node.attrs;
        return (0, mithril_1.default)('.track-content', {
            onmousemove: (e) => {
                attrs.track.onMouseMove({ x: e.layerX - css_constants_1.TRACK_SHELL_WIDTH, y: e.layerY });
                globals_1.globals.rafScheduler.scheduleRedraw();
            },
            onmouseout: () => {
                attrs.track.onMouseOut();
                globals_1.globals.rafScheduler.scheduleRedraw();
            },
            onmousedown: (e) => {
                this.mouseDownX = e.layerX;
                this.mouseDownY = e.layerY;
            },
            onmouseup: (e) => {
                if (this.mouseDownX === undefined ||
                    this.mouseDownY === undefined) {
                    return;
                }
                if (Math.abs(e.layerX - this.mouseDownX) > 1 ||
                    Math.abs(e.layerY - this.mouseDownY) > 1) {
                    this.selectionOccurred = true;
                }
                this.mouseDownX = undefined;
                this.mouseDownY = undefined;
            },
            onclick: (e) => {
                // This click event occurs after any selection mouse up/drag events
                // so we have to look if the mouse moved during this click to know
                // if a selection occurred.
                if (this.selectionOccurred) {
                    this.selectionOccurred = false;
                    return;
                }
                // Returns true if something was selected, so stop propagation.
                if (attrs.track.onMouseClick({ x: e.layerX - css_constants_1.TRACK_SHELL_WIDTH, y: e.layerY })) {
                    e.stopPropagation();
                }
                globals_1.globals.rafScheduler.scheduleRedraw();
            },
        }, node.children);
    }
}
exports.TrackContent = TrackContent;
class TrackComponent {
    view({ attrs }) {
        // TODO(hjd): The min height below must match the track_shell_title
        // max height in common.scss so we should read it from CSS to avoid
        // them going out of sync.
        return (0, mithril_1.default)('.track', {
            style: {
                height: `${Math.max(18, attrs.track.getHeight())}px`,
            },
            id: 'track_' + attrs.trackState.id,
        }, [
            (0, mithril_1.default)(TrackShell, {
                track: attrs.track,
                trackState: attrs.trackState,
                pinnedCopy: attrs.pinnedCopy
            }),
            (0, mithril_1.default)(TrackContent, { track: attrs.track }),
        ]);
    }
    oncreate({ attrs }) {
        if (globals_1.globals.frontendLocalState.scrollToTrackId === attrs.trackState.id) {
            (0, scroll_helper_1.verticalScrollToTrack)(attrs.trackState.id);
            globals_1.globals.frontendLocalState.scrollToTrackId = undefined;
        }
    }
}
class TrackButton {
    view({ attrs }) {
        return (0, mithril_1.default)('i.track-button', {
            class: [
                (attrs.showButton ? 'show' : ''),
                (attrs.fullHeight ? 'full-height' : ''),
                (attrs.filledIcon ? 'material-icons-filled' : 'material-icons'),
                (attrs.disabled) ? 'disabled' : '',
            ].filter(Boolean).join(' '),
            onclick: attrs.disabled ? null : attrs.action,
            title: attrs.tooltip,
        }, attrs.i);
    }
}
exports.TrackButton = TrackButton;
class TrackPanel extends panel_1.Panel {
    constructor(vnode) {
        super();
        const trackId = vnode.attrs.id;
        this.pinnedCopy = vnode.attrs.pinnedCopy;
        const trackState = globals_1.globals.state.tracks[trackId];
        if (trackState === undefined) {
            return;
        }
        const engine = globals_1.globals.engines.get(trackState.engineId);
        if (engine === undefined) {
            return;
        }
        const trackCreator = track_registry_1.trackRegistry.get(trackState.kind);
        this.track = trackCreator.create({ trackId, engine });
        this.trackState = trackState;
    }
    view() {
        if (this.track === undefined || this.trackState === undefined) {
            return (0, mithril_1.default)('div', 'No such track');
        }
        return (0, mithril_1.default)(TrackComponent, {
            trackState: this.trackState,
            track: this.track,
            pinnedCopy: this.pinnedCopy
        });
    }
    oncreate() {
        if (this.track !== undefined) {
            this.track.onFullRedraw();
        }
    }
    onupdate() {
        if (this.track !== undefined) {
            this.track.onFullRedraw();
        }
    }
    onremove() {
        if (this.track !== undefined) {
            this.track.onDestroy();
            this.track = undefined;
        }
    }
    highlightIfTrackSelected(ctx, size) {
        const { visibleTimeScale } = globals_1.globals.frontendLocalState;
        const selection = globals_1.globals.state.currentSelection;
        const trackState = this.trackState;
        if (!selection || selection.kind !== 'AREA' || trackState === undefined) {
            return;
        }
        const selectedArea = globals_1.globals.state.areas[selection.areaId];
        const selectedAreaDuration = selectedArea.end - selectedArea.start;
        if (selectedArea.tracks.includes(trackState.id)) {
            ctx.fillStyle = (0, css_constants_1.getCssStr)('--selection-fill-color');
            ctx.fillRect(visibleTimeScale.tpTimeToPx(selectedArea.start) + css_constants_1.TRACK_SHELL_WIDTH, 0, visibleTimeScale.durationToPx(selectedAreaDuration), size.height);
        }
    }
    renderCanvas(ctx, size) {
        ctx.save();
        // If we have vsync data, render columns under the track and
        // under the grid lines painted next
        const vsync = (0, vsync_helper_1.getActiveVsyncData)();
        if (vsync) {
            ctx.save();
            ctx.translate(css_constants_1.TRACK_SHELL_WIDTH, 0);
            (0, vsync_helper_1.renderVsyncColumns)(ctx, size.height, vsync);
            ctx.restore();
        }
        (0, gridline_helper_1.drawGridLines)(ctx, size.width, size.height);
        ctx.translate(css_constants_1.TRACK_SHELL_WIDTH, 0);
        if (this.track !== undefined) {
            this.track.render(ctx);
        }
        ctx.restore();
        this.highlightIfTrackSelected(ctx, size);
        const { visibleTimeScale } = globals_1.globals.frontendLocalState;
        // Draw vertical line when hovering on the notes panel.
        if (globals_1.globals.state.hoveredNoteTimestamp !== -1n) {
            (0, vertical_line_helper_1.drawVerticalLineAtTime)(ctx, visibleTimeScale, globals_1.globals.state.hoveredNoteTimestamp, size.height, `#aaa`);
        }
        if (globals_1.globals.state.hoverCursorTimestamp !== -1n) {
            (0, vertical_line_helper_1.drawVerticalLineAtTime)(ctx, visibleTimeScale, globals_1.globals.state.hoverCursorTimestamp, size.height, `#344596`);
        }
        if (globals_1.globals.state.currentSelection !== null) {
            if (globals_1.globals.state.currentSelection.kind === 'SLICE' &&
                globals_1.globals.sliceDetails.wakeupTs !== undefined) {
                (0, vertical_line_helper_1.drawVerticalLineAtTime)(ctx, visibleTimeScale, globals_1.globals.sliceDetails.wakeupTs, size.height, (0, css_constants_1.getCssStr)('--main-foreground-color'));
            }
        }
        // All marked areas should have semi-transparent vertical lines
        // marking the start and end.
        for (const note of Object.values(globals_1.globals.state.notes)) {
            if (note.noteType === 'AREA') {
                const transparentNoteColor = 'rgba(' + color_convert_1.hex.rgb(note.color.substr(1)).toString() + ', 0.65)';
                (0, vertical_line_helper_1.drawVerticalLineAtTime)(ctx, visibleTimeScale, globals_1.globals.state.areas[note.areaId].start, size.height, transparentNoteColor, 1);
                (0, vertical_line_helper_1.drawVerticalLineAtTime)(ctx, visibleTimeScale, globals_1.globals.state.areas[note.areaId].end, size.height, transparentNoteColor, 1);
            }
            else if (note.noteType === 'DEFAULT') {
                (0, vertical_line_helper_1.drawVerticalLineAtTime)(ctx, visibleTimeScale, note.timestamp, size.height, note.color);
            }
        }
    }
    getSliceRect(tStart, tDur, depth) {
        if (this.track === undefined) {
            return undefined;
        }
        return this.track.getSliceRect(tStart, tDur, depth);
    }
}
exports.TrackPanel = TrackPanel;
//# sourceMappingURL=track_panel.js.map
```

jcortell68 commented:
So, the drag stopping seems more reasonable now. One thing I noticed is that if the selection/details pane is open (e.g., click on a slice in one of the tracks), the scrolling doesn't stop at the top of the selection pane (at the bottom of the track viewport). It should, IMO.

jcortell68 commented:
> From my opinion killing resizing after the mouse leaves the timeline could cause situations where users may not be able to resize tracks or track groups as large as they want.

I've been thinking about this. I _think_ I know what scenario you're referring to. 

1. Open a profile
2. Scroll to the very bottom
3. Try to make the last track in the timeline taller. You can't, since the bottom edge of that last track is the bottom of the timeline area. If that's the limit for the dragging, then you're kind of screwed.

Here's how I think we could solve that dilemma. The track area shouldn't end at the bottom of the last track, I.e., in the example above, you should be able to continue scrolling for a good 400-500 pixels and black should appear below the last track. The trailing "emptiness" is not a novel concept. Maximize the Sokatoa window and then do a Collapse All on the tracks. You'll see quite a bit of emptiness past the last track. That's what I'm saying should _always_ be at the bottom of the tracks. Then, you'd always be able to enlarge the last track. You just have to scroll to the end and you should have enough room to drag the last tracks border down.

ALevansSamsung committed:
Stop resize when mouse enters selection panel

ALevansSamsung committed:
add 500px of empty space to bottom of timeline scroll view

ALevansSamsung commented:
New versions of track_group_panel.js and track_panel.js
[built_2.zip](https://github.com/user-attachments/files/15995250/built_2.zip)

For the .scss file change.  This css rule can be added to any .css file to get a similar effect.
```
.perfetto .scrolling-panel-container>.panels>*:last-child {
    margin-bottom: 500px;
}
```

jcortell68 commented:
@ALevansSamsung Version two is looking good! 

As part of testing it, I noticed another buggy behavior. But this is pre-existing, so I don't know if you want to handle it as part of this PR or if you want me to create a dedicated one.

In some cases....during a resize, the mouse position and the border being dragged become separated, and the divergence grows  the further down you drag. To reproduce:

1. Collapse All
2. Drag the bottom border of the `Processes` track-group and notice this works as expected. The mouse and the border stay together, even when you drag it all the way to the bottom
3. Expand the Processes group and now resize the `surfaceflinger` group. Start by making it only a little bit taller. So far so good. Keep going, you'll notice some distance between mouse and border forming. Now drag all the way to the bottom, you'll notice a significant distance between mouse and border. Basically, I should be able to make it taller, but I can't because the mouse gets to the bottom of the timeline, but the border is lagging behind.

jcortell68 commented:
BTW, for future reference, here's how I was able to apply the js/css changes locally on Windows:

1. Replace the js files in `external\perfetto\ui\out\tsc` with the ones provided
2. Enhance/modify  `external\perfetto\ui\css\perfetto.css` to have the .scss changes provided. Note that the file has a dense format and every entry is qualified with `.perfetto`. So, e.g., for this particular change, at the bottom of the file I added
```
.perfetto .scrolling-panel-container > .panels > *:last-child {
    margin-bottom: 500px;
 }
 ```
 If instead of an additional selector, you had to modify an existing selector, find and modify the selector in the dense listing
 
 Of course, if the developer provided a `perfetto.css`, life is easier. Just replace your file with the provided one.
 
 3. The next steps depend on whether you have an active `yarn watch`.

If you do have `yarn watch` running:
* Ctrl+C the yarn watch
* `yarn perfetto-deps`
* `yarn watch`
* wait. How long? 1-2 minutes. If you want to know exactly how long, run `python scripts/yarnwatch_latency.py`. When that completes, the yarn watch has webpacked the modified files
* `yarn electron start:asis` to launch Sokatoa and see the changes

If you do **not** have `yarn watch` running:
* `yarn perfetto-deps`
* `yarn electron-start` to webpack the modified files and launch Sokatoa to see the changes

ALevansSamsung commented:
> In some cases....during a resize, the mouse position and the border being dragged become separated, and the divergence grows the further down you drag. To reproduce:
> 
> 1. Collapse All
> 2. Drag the bottom border of the `Processes` track-group and notice this works as expected. The mouse and the border stay together, even when you drag it all the way to the bottom
> 3. Expand the Processes group and now resize the `surfaceflinger` group. Start by making it only a little bit taller. So far so good. Keep going, you'll notice some distance between mouse and border forming. Now drag all the way to the bottom, you'll notice a significant distance between mouse and border. Basically, I should be able to make it taller, but I can't because the mouse gets to the bottom of the timeline, but the border is lagging behind.

Each Track currently specifies their own height.  That means that we can't easily get the 1 for 1 height that the track is going to be after the scaleFactor changes.  Some tracks have constant margins in their height calculations that are not scaled.  This causes the mouse to slowly pull away from the position of the tracks bottom border.

I see a few ways we could fix this:

1.  Be more transparent with the fact that the user is scaling as a multiplier rather than pulling the bottom border
2.  Change scaleFactor to act as a height override in the Track API (would require changing all track renderers) 
3.  Remove margins and any variables causing tracks to not have a 1:1 scaling from track renders (would require changing less than all track renderers)

jcortell68 commented:
> I see a few ways we could fix this:

I don't think (3) is viable. I'm not sure what (1) would entail, but at the end of the day, it would still prove to be unintuitive (even if transparent). I think (2) is the way to go, even if it means changing each renderer. In the end, there aren't hundreds of them. A couple of dozen at most. And I think the adjustments would be small enough that it won't make rebasing to a newer perfetto tons more difficult. It's going to already be a nightmare. What's a little more pain? IMO, the current behavior feels buggy and I think it would be good to fix it. But maybe something @wpaul-samsung can help prioritize for us.

jcortell68 commented:
BTW, like I said this is pre-existing behavior, unrelated to the ask in this ticket. If I were you, I'd wrap this one up and we can handle this other thing separately. I'm going to do a line-by-line review of the changes now and hopefully +1 it shortly.

jcortell68 commented on ui/src/frontend/track_panel.ts:
```diff
@@ -191,56 +191,77 @@ class TrackShell implements m.ClassComponent<TrackShellAttrs> {
-            this.attrs.trackState.scaleFactor = 1;
-          } else {
-            this.attrs.trackState.scaleFactor = newMultiplier;
+    if (e.currentTarget instanceof HTMLElement) {
```
On the surface, the logic/code for this (a track) looks the same as for a track group.  And the code involved is non-trivial. I.e., seems like a good amount of code duplication. Could you extract this into some common code that both panel classes use?

jcortell68 commented on ui/src/frontend/track_group_panel.ts:
```diff
@@ -101,55 +101,77 @@ export class TrackGroupPanel extends Panel<Attrs> {
+          mouseY = evMove.clientY;
+          if (this.attrs && this.defaultHeight) {
+            const newMultiplier = trackHeight / this.defaultHeight;
+            if (newMultiplier < 1) {
```
style nit: 
```
this.summaryTrackState.scaleFactor = Math.min(1, newMultiplier);
```

Five lines of code down to one without, IMO, any loss in readability.

ALevansSamsung committed:
Consolidate duplicate resize logic

ALevansSamsung committed:
use Math.min

ALevansSamsung commented:
Javascript file changes

[build_3.zip](https://github.com/user-attachments/files/16017801/build_3.zip)

jcortell68 commented:
LGTM. Thanks for the adjustments. Didn't test-drive the latest version of the code, since it's effectively code relocating/condensing.  I assume you tested for regressions. Nice job! +1
